### PR TITLE
Make things generic?

### PR DESCRIPTION
**Pull request by taion**

### CODE REVIEW

File: src/resources/HttpResource.js [hunk 1]
```diff
@@ -1,21 +1,24 @@
 /* @flow */
 
 import type HttpApi, { Args, Data } from '../api/HttpApi';
-import type { Context } from '../types/NodeType';
 import urlJoin from '../utils/urlJoin';
 import Resource from './Resource';
 
+type HttpContext = {
```
Comment (taion) [inline src/resources/HttpResource.js]: this should probably be generic

**Comment by jquense** [comment]
I can't get this to work when actually trying to define a NodeType with a HttpResource :/

**Comment by taion** [comment]
Do you have example code up? I can try to debug a bit

**Comment by jquense** [comment]
try this:
```
/* @flow */

import { GraphQLString, GraphQLBoolean, GraphQLFloat } from 'graphql';

import createNodeType from './types/NodeType';
import HttpResource from './resources/HttpResource';

const { NodeType } = createNodeType({});
// const NodeT: NodeType<HttpResource> = NodeType;

class DeviceResource extends HttpResource<*> {}

const Device = new NodeType({
  name: 'Device',

  fields: () => ({
    name: { type: GraphQLString, from: 'deviceName' },
  }),

  Resource: DeviceResource,
});
```

**Comment by taion** [comment]
Okay, this is a disaster. I've spent an hour banging my ahead against this and now I have more errors than I started with.

**Comment by jquense** [comment]
Hahaha flow. I don't know why you can't specify the generic during a function call, like a real language

**Comment by taion** [comment]
Okay, more time down the drain... I have an _awful_ solution.

I think the best way to go is to not do this awful thing and just return a factory method. Ugh.

**Comment by taion** [comment]
half an hour later i have a real solution

**Comment by taion** [comment]
sort of...

**Comment by taion** [comment]
the issue isn't about specifying the function call... the issue is that the `NodeType` we return is a partially bound generic type, so you just can't express something like `Class<NodeType<? extends Resource<BOUND_CONTEXT>>>`... like that's just not a concept that the type system allows expressing... for one thing that class doesn't even exist outside the closure

**Comment by taion** [comment]
we could probably make the `createNodeType` thing build an internal `NodeType` class and remove some of the internal hackiness. we wouldn't be able to extend this thing any more, but actually the only case where we extend it is to add a few extra fields, and we can do that just as easily by just spreading those fields in

overall this is a huge mess, but it's also sort of ok?

**Comment by jquense** [comment]
Is there a ton of value in typing Context here to begin with? Maybe it's simpler just to make it an `any` and then type it back in the Resources themselves? Like the places it's created are super scoped to the one entry file, and for other cases (like subscriptions) we have a pretty good handle on it

**Comment by taion** [comment]
Good point. There's no real value to typing context here.

**Pull request closed without merging**